### PR TITLE
fix(@desktop/sections): make ctrl+k display chats from communities

### PR DESCRIPTION
**Pull request by osmaczko**
Closes: #4059

**Updated version:**

https://user-images.githubusercontent.com/33099791/152864269-65c775d6-f23e-4085-b5ec-1bb3f9ee8091.mp4

Previous version:

https://user-images.githubusercontent.com/33099791/152536507-a04580be-f8a1-4ee8-b33f-31ebf5a4784b.mp4

### CODE REVIEW

File: src/app/modules/main/module.nim [hunk 1]
```diff
@@ -50,7 +51,7 @@ import ../../core/eventemitter
 
 export io_interface
 
-type
```
Comment (saledjenic) [inline src/app/modules/main/module.nim:53]: @osmaczko first of all this PR is very hard to follow due to lot of changes like this one. So please revert all that and apply changes only you did for this particular task.

File: src/app/modules/main/view.nim [hunk 2]
```diff
@@ -111,12 +114,18 @@ QtObject:
...
       return newQVariant()
-    
+
     return communityVariant
 
   proc getChatSectionModule*(self: View): QVariant {.slot.} =
     return self.delegate.getChatSectionModule()
 
+  proc getAllChatsModel*(self: View): QVariant {.slot.} =
+    return self.delegate.getAllChatsModel()
```
Comment (jrainville) [inline src/app/modules/main/view.nim:124]: This would be better as a QtProperty. Example:
```nim
  QtProperty[QVariant] allChatsModel:
    read = getAllChatsModel
    notify = allChatsModelChanged # this signal would need to be added
```

File: src/app/modules/main/chat_section/model.nim [hunk 3]
```diff
@@ -23,7 +23,7 @@ type
 QtObject:
   type
     Model* = ref object of QAbstractListModel
-      items: seq[Item]
+      items*: seq[Item]
```
Comment (saledjenic) [inline src/app/modules/main/chat_section/model.nim:26]: @osmaczko exposing private list of items this way somehow breaks the rules we set for the new architecture. That may allow someone to unintentionally set or update the list (`items`) without emitting appropriate signal and bring some inconsistencies here. Also exposing private members is hard to follow/maintain in a big project. So please revert this.

File: src/app/modules/main/chat_section/module.nim [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import NimQml, Tables, chronicles, json, sequtils
 import io_interface
+import model as chats_model
```
Comment (saledjenic) [inline src/app/modules/main/chat_section/module.nim:3]: @osmaczko this `model` was already imported in the line num 5. Please revert this line.

File: ui/app/AppMain.qml [hunk 5]
```diff
@@ -800,26 +802,22 @@ Item {
...
-                return statusSmartIdenticonComponent.createObject(parent, {
-                                                                     imageSource: modelData.identicon,
-                                                                     name: modelData.name
-                                                            });
-            }
-
-            onClicked: function (index, id) {
-                Global.changeAppSectionBySectionType(Constants.appSection.chat)
-                mainModule.getChatSectionModule().setActiveItem(id, "")
+            modelList: rootStore.mainModuleInst.getAllChatsModel()
```
Comment (jrainville) [inline ui/app/AppMain.qml:805]: If possible, you can expose the `allChatsModel` from the store directly, instead of passing by `mainModuleInst`. So something like this:
```qml
 modelList: rootStore.allChatsModel 
```

And in the store you have
```qml
allChatsModel : mainModuleInst.allChatsModel // This will only work when the QtProperty will be exposed
```

File: ui/imports/shared/status/StatusInputListPopup.qml [hunk 6]
```diff
@@ -130,46 +140,52 @@ Popup {
...
 
         delegate: Rectangle {
             property variant myData: typeof modelData === "undefined" ? model : modelData
-            property string myText: popup.getText(myData)
+            property string myText: root.getText(myData)
             id: rectangle
             visible: searchBox.text === "" || myText.includes(searchBox.text)
-            color: listView.currentIndex === index ? Style.current.backgroundHover : Style.current.transparent
+            color: listView.currentIndex
+                   === index ? Style.current.backgroundHover : Style.current.transparent
```
Comment (jrainville) [inline ui/imports/shared/status/StatusInputListPopup.qml:160]: Just aesthetic, but it would be prettier if the line break was after `?`, that way the condition stays whole

**Review by jrainville — CHANGES_REQUESTED** [review]
Nice work. 

I have to request changes, because some of the code here doesn't follow the architecture, but it's easy fixes, like moving the Model to the View instead.

Let me know if anything is unclear in my comments.

File: ui/app/AppMain.qml [hunk 7]
```diff
@@ -800,26 +802,22 @@ Item {
             width: 350
             x: parent.width / 2 - width / 2
             y: parent.height / 2 - height / 2
-            // TODO improve this to work with community Chats as well
-            modelList: mainModule.getChatSectionModule().model
```
Comment (saledjenic) [inline ui/app/AppMain.qml:804]: @osmaczko I connected a comment here, but it's about `StatusInputListPopup`, so this `StatusInputListPopup` component is not something we're proud of, would be good to check with @PascalPrecht is there already any convenient component on the `StatusQ` side which may replace this one.

For this component we should just forward a model with expected props exposed and it should know how to do the rest, and it should have just a signal e.g. "itemClicked(id)" to notify outside world what's happened in there.

If we don't have such component we may improve this one a bit in that direction.

File: src/app/modules/main/module.nim [hunk 8]
```diff
@@ -66,6 +67,8 @@ type
     communitiesModule: communities_module.AccessInterface
     appSearchModule: app_search_module.AccessInterface
     nodeSectionModule: node_section_module.AccessInterface
+    allChatsModel: chat_model.Model
```
Comment (saledjenic) [inline src/app/modules/main/module.nim:70]: @osmaczko `allChatsModel` and  `allChatsModelVariant` should be part of the `view` class, cause `view` is our adapter for interaction with the UI (qml). So please update that accordingly.

**Review by saledjenic — CHANGES_REQUESTED** [review]
@osmaczko you did the task and it works except in the case if community has a category/categories which has/have channel/s and it doesn't display user's identicon/image. But the thing here is about our new architecture and some rules we're trying to apply. Please read a comments bellow and here I am going to describe how it would be and please update your current changes accordingly.

I see that you used `model` form the `chat_content` section which is maybe too heavy for what you need here (so you could create a new one which will have only id, color, image, name), but anyway if you want to use it here, that means that model is shared among two modules and it should be moved to the `shared_models` directory and used on both places from there.

Apart of that there are 2 options how you should proceed:

- Since our services are the source of truth for us, you can use them e.g. `self.controller.getJoinedCommunities()` to get all joined communities and then to follow `buildChatUI` and `buildCommunityUI` procs from `chat section` module and how we're building list of channels for Chat and for each Community section and apply the similar for your need. So you can use this approach if you decide to have your own model for this purpose.

- Another acceptable approach is to do similar what you did, but to add method to `chats model` which will return list of items (which are actually chats/channels) and then use them here.

In both cases model you're building for this purpose must be recreated every time the app popup is raised (every time one does `ctrl+k`).

That's it so far, lets change that firstly.

File: src/app/modules/main/module.nim (hunk 1 above)
Comment (jrainville) [reply src/app/modules/main/module.nim:53]: Reverting is probably not necessary, but sure, maybe we can do the clean up in another PR next time

File: ui/app/AppMain.qml (hunk 7 above)
Comment (PascalPrecht) [reply ui/app/AppMain.qml:804]: Hi!

Missing the context here. What is a `StatusInputList`?

Comment (saledjenic) [reply ui/app/AppMain.qml:804]: @PascalPrecht it's a simple poup actually consists of two parts, top section which is a search box and list of items as a section below that search box. Typing a word in search box you're filter out the list and a click on the item from that list should emit a signal with id of the clicked item e.g. "itemClicked(item-id)". Do we have something similar on the `StatusQ` side. This is a current implementation we're using [here](https://github.com/status-im/status-desktop/blob/master/ui/imports/shared/status/StatusInputListPopup.qml). And [here](https://github.com/status-im/status-desktop/pull/4727#issue-840258910) is a video displaying its use cases.

Comment (PascalPrecht) [reply ui/app/AppMain.qml:804]: Ah I see!
No we don't have such a thing in StatusQ, however, it sounds like this is something that can/should be composed of existing StatusQ components (`StatusInput`, `StatusListItem` etc)

File: src/app/modules/main/chat_section/model.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (osmaczko) [reply src/app/modules/main/chat_section/model.nim]: I saw this approach somewhere in the code before, that's why I used it.
```
$ grep -rn "items\*: seq" src/
src/app/modules/main/profile_section/sync/model.nim:11:    items*: seq[Item]
src/app/modules/main/profile_section/devices/model.nim:13:    items*: seq[Item]
src/app/modules/main/communities/models/pending_request_model.nim:15:    items*: seq[PendingRequestItem]
src/app/modules/shared_models/contacts_model.nim:16:    items*: seq[Item]
```

But sure, I agree we should keep it private to preserve encapsulation.

File: ui/imports/shared/status/StatusInputListPopup.qml (hunk 6 above)
Comment (osmaczko) [reply ui/imports/shared/status/StatusInputListPopup.qml]: This change was reverted.

File: src/app/modules/main/chat_section/model.nim (hunk 3 above)
Comment (osmaczko) [reply src/app/modules/main/chat_section/model.nim]: https://github.com/status-im/status-desktop/pull/4727#discussion_r801024464

File: src/app/modules/main/module.nim (hunk 1 above)
Comment (osmaczko) [reply src/app/modules/main/module.nim]: It's "trim trailing whitespace on file save" IDE feature. Reverted.

I'll create separate PR removing trailing whitespaces across all sources.

File: ui/app/AppMain.qml (hunk 7 above)
Comment (osmaczko) [reply ui/app/AppMain.qml:804]: Please see updated version which uses both StatusInput and StatusListItem -> https://github.com/status-im/status-desktop/pull/4727#issue-840258910

@PascalPrecht I've introduced new component for that: `ui/imports/shared/status/StatusSearchListPopup.qml`. I guess it should go to StatusQ, shouldn't it?

**Comment by osmaczko** [comment]
@saledjenic thanks for the review!

> * in case if community has a category/categories which has/have channel/s they are not displayed in the popup list (also category itself should be omitted)
> * in case of chats, from the Chat section user's identicon/image is not displayed

Good catches, both are fixed now.

> I see that you used `model` form the `chat_content` section which is maybe too heavy for what you need here (so you could create a new one which will have only id, color, image, name)

I have indeed created a new model as you suggested, chat_search_model (of chat_search_items). I couldn't come up with better names..

> * Another acceptable approach is to do similar what you did, but to add method to `chats model` which will return list of items (which are actually chats/channels) and then use them here to build a model according to the needs we have here.

I prefer this approach because ctrl+k should be fast and should be consistent with the data we already show in the application. The other approach you suggested may be too heavy in terms of time as it needs to call status-go to recreate chat models and also may lead to inconsistency (e.g. if there is a chat that wasn't fetched before when constructing chat models due to some status-go bug)

> In both cases model you are building for this purpose must be recreated every time the app popup is raised (every time one does `ctrl+k`).

Yes, I know. It is (and was) done like that. See: `rebuildChatSearchModel` method.

File: src/app/modules/main/chat_section/module.nim (hunk 4 above)
Comment (osmaczko) [reply src/app/modules/main/chat_section/module.nim]: I removed `model` from line num 5, otherwise, there is an ambiguity error with `contacts_model.Model`

**Comment by PascalPrecht** [comment]
@osmaczko The cursor shold turn into a pointer when hovering any of those channel results

File: src/app/modules/main/view.nim [hunk 9]
```diff
@@ -54,6 +61,19 @@ QtObject:
     read = getModel
     notify = modelChanged
 
+  proc chatSearchModelChanged*(self: View) {.signal.}
+
+  proc getChatSearchModel(self: View): QVariant {.slot.} =
+    return self.chatSearchModelVariant
+
+  proc rebuildChatSearchModel*(self: View) {.slot.} =
+    let chatSearchItems = self.delegate.getChatSearchItems()
```
Comment (saledjenic) [inline src/app/modules/main/view.nim:70]: @osmaczko lets move this logic to `module.nim` so here you will have just this:
```
proc rebuildChatSearchModel*(self: View) {.slot.} =
    self.delegate.rebuildChatSearchModel()
```

**Comment by osmaczko** [comment]
> @osmaczko The cursor shold turn into a pointer when hovering any of those channel results

@PascalPrecht done

https://user-images.githubusercontent.com/33099791/153052311-8d97f8db-b300-437f-901d-e4f164729671.mp4

File: src/app/modules/main/module.nim [hunk 10]
```diff
@@ -508,6 +511,27 @@ method getCommunitySectionModule*[T](self: Module[T], communityId: string): QVar
 
   return self.communitySectionsModule[communityId].getModuleAsVariant()
 
+method rebuildChatSearchModel*[T](self: Module[T]) =
+  let transformItem = proc(item: chat_section_base_item.BaseItem, sectionId, sectionName: string): chat_search_item.Item =
+    result = chat_search_item.initItem(item.id(), item.name(), item.color(), item.icon(), sectionId, sectionName)
```
Comment (jrainville) [inline src/app/modules/main/module.nim:516]: You can make `transformItem` an actual proc if you want. Putting a proc inside another proc like that is not something we do often

File: src/app/modules/main/module.nim [hunk 11]
```diff
@@ -508,6 +511,27 @@ method getCommunitySectionModule*[T](self: Module[T], communityId: string): QVar
...
+
+  let transform = proc(items: seq[chat_section_item.Item], sectionId, sectionName: string): seq[chat_search_item.Item] =
```
Comment (jrainville) [inline src/app/modules/main/module.nim:518]: Same here

File: ui/imports/shared/status/StatusSearchListPopup.qml [hunk 12]
```diff
@@ -0,0 +1,166 @@
+import QtQuick 2.14
```
Comment (jrainville) [inline ui/imports/shared/status/StatusSearchListPopup.qml:1]: Can we use this new component for the other place that `StatusInputListPopup` was used? It can be done in another PR, but it would be nice to not keep both components since they basically do the same thing

**Review by jrainville — APPROVED** [review]
Good job

**Review by saledjenic — APPROVED** [review]

File: src/app/modules/main/module.nim (hunk 10 above)
Comment (osmaczko) [reply src/app/modules/main/module.nim:516]: In this case, I would like to keep the procedure inside `rebuildChatSearchModel`, because If I move it outside it will lose its context. It's also a few-liner so the readability still stays on the high level.

File: ui/imports/shared/status/StatusSearchListPopup.qml (hunk 12 above)
Comment (osmaczko) [reply ui/imports/shared/status/StatusSearchListPopup.qml:1]: Good point, I will try to deliver it in the next PR, altogether with `StatusInputListPopup` removal.

**Review by PascalPrecht — APPROVED** [review]
